src/multiplexer/EventHandler.cpp

```diff
@@ -196,26 +206,45 @@ void EventHandler::handleResponse(int clientFd) {
 	changeToRead(clientFd);
 }
 
-void EventHandler::epollLoop(Server & s) {
+void EventHandler::checkCompleteCGIProcesses(void) {
+	std::map<int, CGIInfo*>::iterator it;
+	for (it = _cgiManager._cgiProcesses.begin();
+			it != _cgiManager._cgiProcesses.end();
+			++it) {
+		if (it->second->isFinished) {
+			CGIInfo *info = it->second;
+			int clientFd = info->clientFd;
+			_clients[clientFd]->_responseBuffer = info->output;
+			deleteFromEpoll(info->pipeFd);
+			_cgiManager.deleteFromCGI(info->pipeFd);
+			changeToWrite(clientFd);
+		}
+	}
+}
+
+void EventHandler::epollLoop(void) {
 	struct epoll_event eventQueue[MAX_EVENTS];
-	int serverFd = s.getSockFd();
 
 	while (1) {
-		// change from -1 after bug is fixed
-		int numEvents = epoll_wait(_epollFd, eventQueue, MAX_EVENTS, -1);
+		int numEvents = epoll_wait(_epollFd, eventQueue, MAX_EVENTS, 0);
 		if (numEvents == -1) {
 			throw epollWaitFailure();
 		}
-		if (numEvents != 0)
-			std::cout << "Num events: " << numEvents << "\n";
 		for (int i = 0; i < numEvents; ++i) {
-			if (eventQueue[i].data.fd == serverFd) {
-				handleNewConnection(s);
+			// New connection from server
```

Comment:
for this whole part i'd really recommend changing the epoll structure to take in a struct that has as it's member:
a epoll type (server, client, cgi),
a void ptr to store the data (so the ptr of the server, client or cgi),

and store it inside a map to with a key/value pair of (FD, custom_epoll_struct)

it would make it a lot easier to query which type it is and juggle the data without getting lost

Reply:
good point, will work on this today